java/src/main/java/com/google/protobuf/ZeroCopyEncoder.java

```diff
@@ -0,0 +1,647 @@
+                buffer.arrayOffset() + buffer.position(),
+                bufferSize - buffer.position());
+        buffer.position(position);
+        charPos += charsToWrite;
+      }
+    }
+  }
+
+  /**
+   * Compute the number of bytes that would be needed to encode a varint. {@code value} is treated
+   * as unsigned, so it won't be sign-extended if negative.
+   */
+  private static int computeRawVarint32Size(final int value) {
+    if ((value & (0xffffffff << 7)) == 0) return 1;
+    if ((value & (0xffffffff << 14)) == 0) return 2;
+    if ((value & (0xffffffff << 21)) == 0) return 3;
+    if ((value & (0xffffffff << 28)) == 0) return 4;
+    return 5;
+  }
+
+  @Override
+  public void writeGroupNoTag(MessageLite value) throws IOException {
+    value.writeTo(this);
+  }
+
+  @Override
+  public void writeMessageNoTag(MessageLite value) throws IOException {
+    writeRawVarint32(value.getSerializedSize());
+    value.writeTo(this);
+  }
+
+  @Override
+  public void writeBytesNoTag(ByteString value) throws IOException {
+    writeRawVarint32(value.size());
+    writeRawBytes(value);
+  }
+
+  @Override
+  public void writeByteArrayNoTag(byte[] value) throws IOException {
+    writeRawVarint32(value.length);
+    writeRawBytes(value);
+  }
+
+  @Override
+  public void writeByteArrayNoTag(byte[] value, int offset, int length) throws IOException {
+    writeRawVarint32(length);
+    writeRawBytes(value, offset, length);
+  }
+
+  @Override
+  public void writeUInt32NoTag(int value) throws IOException {
+    writeRawVarint32(value);
+  }
+
+  @Override
+  public void writeEnumNoTag(int value) throws IOException {
+    writeInt32NoTag(value);
+  }
+
+  @Override
+  public void writeSFixed32NoTag(int value) throws IOException {
+    writeRawLittleEndian32(value);
+  }
+
+  @Override
+  public void writeSFixed64NoTag(long value) throws IOException {
+    writeRawLittleEndian64(value);
+  }
+
+  @Override
+  public void writeSInt32NoTag(int value) throws IOException {
+    writeRawVarint32(encodeZigZag32(value));
+  }
+
+  @Override
+  public void writeSInt64NoTag(long value) throws IOException {
+    writeRawVarint64(encodeZigZag64(value));
+  }
+
+  @Override
+  public void flush() throws IOException {
+    if (position() == 0) {
+      // Nothing to flush.
+      return;
+    }
+
+    buffer.flip();
+    int limit = buffer.limit();
+    if (limit < SMALL_BUFFER || (bufferSize / limit) < COPY_THRESHOLD) {
+      // Either not much data to onDataEncoded or the buffer is not very full ... copy the buffer.
+      byte[] bytes = new byte[limit];
+      buffer.get(bytes);
+      buffer.position(0);
+      handler.onDataEncoded(bytes, 0, limit);
+    } else {
+      // Use the current buffer and create a new buffer on the next onDataEncoded.
+      ByteBuffer buf = buffer;
+      buffer = null;
+      handler.onDataEncoded(buf);
+    }
+  }
+
+  @Override
+  public void writeRawByte(byte value) throws IOException {
+    if (capacity() == 0) {
```

Comment:
I think the corollary here is that there are WAY too many public methods in this class...